Generate: New `Cache` abstraction and Attention Sinks support

tomaarsen opened the pull request:
Closes #26553

Hello!

# What does this PR do?
I had a few hours on Saturday to work up a draft version of the updated KV caching mechanism as discussed in #26553. Ideally, this should allow Attention Sinks (https://github.com/tomaarsen/attention_sinks) / StreamingLLM (https://arxiv.org/abs/2309.17453) to be easily implemented in a third-party or in transformers directly. 

The implementation doesn't work well yet, as the VRAM usage quickly shoots up after generating even just 8 tokens. This is probably some bug that I haven't had time for yet. There's a few other comments that I have on specific sections of code, so I'll write some comments below.

## Goal for this draft
The intention for this draft is to continue discussion about whether this is moving in the right direction, and to determine the scope (e.g. do we want to include this updated `Cache` for all architectures that use KV caching?). 

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

@patrickvonplaten 
@gante
@LysandreJik 
@Guangxuan-Xiao

- Tom Aarsen

tomaarsen committed:
Draft version of new KV Caching

This should allow Attention Sinks (https://github.com/tomaarsen/attention_sinks)
/ StreamingLLM (https://arxiv.org/abs/2309.17453) to be easily implemented
in a third-party or in transformers directly

tomaarsen commented on src/transformers/cache_utils.py:
```diff
@@ -0,0 +1,89 @@
+    def __getitem__(self, index: int):
+        return self.cache[self.layer_idx][index]
+
+    def set_layer_idx(self, layer_idx: int) -> None:
+        self.layer_idx = layer_idx
```
We want our `Cache` instance to work both for `past_key_values` and `past_key_value`. This is one of the approaches to make that work - when we want to get the KV cache for a specific layer we first set the layer idx and then slicing works as if we have `past_key_value`.

Other alternatives might require creating some instance for each of the (e.g. 32) layers and then having two-way commmunication between them, which I thought was suboptimal. But I'm not 100% a fan of this approach either, as you now can't use indexing to select layers like you might expect.

gante replied:
Regardless of the class structure we decide to keep, it would be really really great if we could keep the same indexing capabilities as the previous `past_key_values` structure. i.e. the old `past_key_values[a, b, c, d]` would match the new Cache-based `past_key_values[a, b, c, d]`

This would make the transition on the `generation` side much easier -- advanced features like assisted generation or fixed-sized caches rely on indexing, and it would make the transition process much easier 💛

tomaarsen replied:
Makes sense. I'm not sure how `past_key_values` is indexed from the generation mixin as I've only looked at that code a few times, but I'll try to ensure 1-1 parity for indexing. I think this should be feasible, I already have ideas 😄

patrickvonplaten replied:
I would actually not add such a function and also not add a `self.layer_index` attribute to the cache class. Instead we can pass the index directly to `update(...)` so that we can keep the class as easy-to-understand as possible. This has the following advantages:
- The user sees less logic related to the cache in the modeling file (only one `update(...)`) function and instead of `set_layer_index(...)` and `update(...)` => easier to understand
- We keep the cache object more light-weight
- It doesn't really make sense for a cache to be have a layer_index attribute IMO

tomaarsen replied:
I can remove the `layer_idx` from `Cache`, but the layer idx will have to be passed all the way down to the `...Attention` classes instead where `update` is called then, either via `forward` or `__init__`. I was trying to avoid this.
Or otherwise the `Cache` needs to maintain a list of smaller `KVCache` (or some other name) instances, one for each layer, and pass those down as `past_key_value` instead. That would also work, but it's not exactly lightweight either to have two separate classes and e.g. 33 instances just for one cache.

I do agree with your 3 advantages, though

tomaarsen replied:
Oh, I didn't realise you already did this. Yeah, I'm totally fine with that then.

tomaarsen replied:
Tackled in 1129513b3fe3a540bbeebe4fcef8bc5b29903ad8.

tomaarsen commented on src/transformers/models/llama/modeling_llama.py:
```diff
@@ -352,18 +354,24 @@ def forward(
         key_states = key_states.view(bsz, q_len, self.num_key_value_heads, self.head_dim).transpose(1, 2)
         value_states = value_states.view(bsz, q_len, self.num_key_value_heads, self.head_dim).transpose(1, 2)
 
+        if isinstance(past_key_value, Cache):
+            past_key_value.update_pre_rotation(key_states, value_states)
```
I'm very much not a fan of having an `isinstance` conditional every time, but it's probably the most convenient way to keep backwards compatibility.

Also, I'm considering renaming `update_pre_rotation` and `update` to `on_...` much like the `TrainerCallback`. I think that might be more extendible etc.
Note that `update_pre_rotation` only exists to eventually support the Attention Sink Cache.

patrickvonplaten replied:
Same, I think it's fine to just assume that inside the attention layer the `past_key_value` is always of type `Cache` - only when returning `past_key_value` from the forward we have to transform it back no? Intermediate classes are not considered public API (they can't be imported from `__init__.py`). Wdyt @gante ?

tomaarsen replied:
Tackled in 1129513b3fe3a540bbeebe4fcef8bc5b29903ad8.

tomaarsen commented on src/transformers/models/llama/modeling_llama.py:
```diff
@@ -352,18 +354,24 @@ def forward(
+
         kv_seq_len = key_states.shape[-2]
-        if past_key_value is not None:
+        if past_key_value:
```
```suggestion
        if past_key_value:
```
This is not just some "simplification" - I've added a `__bool__` to the cache that checks if it's empty. This should work for both the original `past_key_value` tuple and the new Cache instances.

tomaarsen replied:
I removed this in 1129513b3fe3a540bbeebe4fcef8bc5b29903ad8.

tomaarsen commented on src/transformers/models/llama/modeling_llama.py:
```diff
@@ -931,7 +948,11 @@ def custom_forward(*inputs):
 
             if use_cache:
-                next_decoder_cache += (layer_outputs[2 if output_attentions else 1],)
+                present_key_values = layer_outputs[2 if output_attentions else 1]
+                if isinstance(present_key_values, Cache):
+                    next_decoder_cache = present_key_values
+                else:
+                    next_decoder_cache += (present_key_values,)
```
No need to append `Cache` instances into some tuple.

gante commented on src/transformers/cache_utils.py:
```diff
@@ -0,0 +1,89 @@
+        self.layer_idx = 0
+
+    @abstractmethod
+    def update_pre_rotation(self, key_states: torch.Tensor, value_states: torch.Tensor) -> None:
```
nit: "rotation" in the name implies the usage with Rotary Position Embeddings. I'd pick a more generic name :)

tomaarsen replied:
It's actually related to that - this method exists to be called before position embeddings are rotated for RoPE and ALiBi approaches. This is a requirement for attention sinks, where the keys need to be cached before being rotated.

The name still isn't great though, and it probably shouldn't be an abstract method as it would likely only be used by some `AttentionSinkCache`.

patrickvonplaten replied:
Is there no way to only have a `update(...)` function instead of forcing each cache to have to `update(...)` functions? Don't think that's very clean. Models like Llama/Falcon only support either RoPE or Alibi, can't we just call `update` here at the right position with a comment that it has to be done pre-rotation?

patrickvonplaten replied:
Hmm I see, are there any other designs we can think of? E.g. could we just store the rotated k/v cache for the attention sink cache and then for the next step just back-rotate and forward-rotate it? Wouldn't back-rotate and forward-rotate it be the same as to just take the difference in angle for the next step and have the same result (just brainstorming, haven't looked in-detail here)

tomaarsen replied:
I will be working on this - I think it's a good suggestion to help with simplicity, but I'm not a big fan of rotating 2 or 3 times rather than once.

tomaarsen replied:
Successfully implemented this in ba87d7383a7a4fde292d073cff33133d659f61af.
I ended up doing some math to determine the cos/sin required to back-rotate and forward-rotate in one go, resulting in a solution with only a small performance decrease. However, moving this functionality out of `...Attention.forward` and into the `SinkCache` does mean that I have to investigate how easily this translates to other architectures.

patrickvonplaten commented on src/transformers/cache_utils.py:
```diff
@@ -0,0 +1,89 @@
+    def update(self, key_states, value_states) -> None:
+        pass
+
+    def __getitem__(self, index: int):
+        return self.cache[self.layer_idx][index]
```
```suggestion
```

Don't really think we need such a method to begin with no? I'd keep the class as light-weight as possible

tomaarsen replied:
We do if we want `past_key_values` to remain indexable like it currently is.

patrickvonplaten replied:
I was envisioning a bit of a different design here: 
- 1.) We try to avoid any `isinstance(cache, Cache)` statements in all non-public classes such as `LlamaAttention` - instead we assume that all objects the flow inside the model are of the cache class type
- 2.) To keep backwards compatibility, we always convert the DynamicCache to the tuple of tuples (called `past_key_values`) that is expected to be returned from the forward method. Similarly we convert the `past_key_values` to a cache object at the beginning of forward if that isn't the case already. 
I would also go so far as to add a `use_legacy_cache` flag to the forward methods so that we can bit-by-bit deprecate the tuple of tuples `past_key_values`. Also for our `generate` method we could already set this flag to False to not suffer from any potential decrease in speed.

I don't think we can keep 100% backward compatibility between a `cache` object and `past_key_values` anyways so at the moment we cannot return a cache object.

On the other hand, new caches that we implement such as the attention sinks cache of the "compilable" cache should never be returned as a tuple of tuples and also don't have to be backward compatible with a tuple of tuples.

tomaarsen replied:
Tackled in 1129513b3fe3a540bbeebe4fcef8bc5b29903ad8. See this [comment](https://github.com/huggingface/transformers/pull/26681#issuecomment-1759921654) for a follow-up question.

patrickvonplaten reviewed (commented):
Left some first nits, thanks for kick-starting this effort!

patrickvonplaten commented on src/transformers/cache_utils.py:
```diff
@@ -0,0 +1,89 @@
+        pass
+
+    def update(self, key_states: torch.Tensor, value_states: torch.Tensor) -> None:
+        kv_states = torch.cat([key_states[None, :], value_states[None, :]], dim=0)
```
actually upon a second thought this is probably a bad idea to concat k,v states here because it unnecessarily allocates memory. Maybe it would be better to just keep to dictionaries `key_cache`, `value_cache`? This way we save one `torch.cat(...)` operation. Wdty @gante?

tomaarsen replied:
Tackled in 1129513b3fe3a540bbeebe4fcef8bc5b29903ad8.

patrickvonplaten commented on src/transformers/models/llama/modeling_llama.py:
```diff
@@ -352,18 +354,24 @@ def forward(
         query_states, key_states = apply_rotary_pos_emb(query_states, key_states, cos, sin, position_ids)
 
-        if past_key_value is not None:
+        if past_key_value:
             # reuse k, v, self_attention
             key_states = torch.cat([past_key_value[0], key_states], dim=2)
             value_states = torch.cat([past_key_value[1], value_states], dim=2)
 
```
```suggestion
```

patrickvonplaten commented on src/transformers/models/llama/modeling_llama.py:
```diff
@@ -352,18 +354,24 @@ def forward(
             value_states = torch.cat([past_key_value[1], value_states], dim=2)
 
-        past_key_value = (key_states, value_states) if use_cache else None
+        if isinstance(past_key_value, Cache):
+            past_key_value.update(key_states, value_states)
+        elif past_key_value is not None:
+            past_key_value = (key_states, value_states)
```
```suggestion
       past_key_value.update(key_states, value_states)
```

I was envisioning to just assume that the `past_key_value` is **always** of type `Cache` at this point

patrickvonplaten replied:
This shoudn't be backwards breaking no?

freckletonj commented:
`past_key_values` and caching will be an incredible feature.

Two PRs should be made irrelevant if this merges: https://github.com/huggingface/transformers/pull/25086  https://github.com/huggingface/transformers/pull/17574

patrickvonplaten commented:
> @patrickvonplaten @gante Based on your feedback, I intend to make the following changes (when I have a bit more time).
> 
> 1. move `layer_idx` to `update(key_states, value_states, layer_idx)` rather than storing it as a class attribute on `Cache`. This also involves adding `layer_idx` as class attributes to e.g. `LlamaAttention` and `LlamaDecoderLayer`. This also removes the `set_layer_index` magic from `Cache`.
> 2. convert `past_key_values` to Cache instance (if not already) at the start of LlamaAttention.forward. This avoids all `isinstance(...)` calls in `LlamaAttention`, and removes the need for the black magicky `__bool__`.
> 3. convert back to tuple of tuples when returning if some `use_legacy_cache` flag is True. (should this flag be propagated all the way up to `LlamaModel`?)
> 4. use separate `key_cache` and `value_cache` dicts in the `Cache` instance for efficiency (removes a `torch.cat`) and simplicity.
> 
> Regarding [#26681 (comment)](https://github.com/huggingface/transformers/pull/26681#discussion_r1350368080) I would have to experiment. There might be options, but they'll probably be slower than necessary.

Sounds great!

tomaarsen committed:
Address numerous PR suggestions

1. Move layer_idx from cache to ...Attention. Removes confusing set_layer_idx magic.
2. Always convert past_key_values to Cache instance at the start of ...Attention, removes all other isinstance calls.
3. Remove __bool__ and __getitem__ magic as they're confusing.
4. past_key_values.update(key, value, idx) now returns key, value.
5. Add use_legacy_cache flag, defaults to None, i.e. Falsey. This breaks generate for now, until 1) the cache is used is generate() or 2) use_legacy_cache is defaulted to True in generate() until we change it in another PR.
6. Separate key_cache and value_cache.

Some work is still needed to see if the SinkCache can conveniently be implemented with just one update method.

tomaarsen commented:
Addressed the various comments. Beyond that, I also made `past_key_values.update(key, value, idx)` returns `key, value` as you'd expect. Manual generation (i.e. repeated calling of a `LlamaForCausalLM` instance with the `past_key_values`) works well, I even see a ~2% speedup, but don't quote me on that speedup. `model.generate` doesn't work yet because `use_legacy_flag` defaults to None, i.e. Falsey, and `model.generate` doesn't work with the cache yet. @patrickvonplaten @gante Should we go for:
1. Immediately update `model.generate` to work with Cache instances or,
2. Insert `use_legacy_cache=True` as default for `model.generate` until this can be removed in some later PR?

This is all under the assumption that the PR is heading in the right direction 😄 

As a heads up, the sink cache does not work yet. I still need to do some experiments to see if I can store rotated keys and back-rotate + forward-rotate them when the cached keys are requested, rather than storing non-rotated keys. For reference, storing non-rotated keys would require (additional) changes on the side of the `modeling_....py` files, but it would be most computationally efficient. That is what I'll work on next. If it has severe impacts on generation speed, then I will reintroduce a e.g. `update_per_rotate` method to cache the non-rotated keys.

- Tom Aarsen

gante commented:
Hey @tomaarsen 👋 

Re `generate` compatibility: usually, I'm pro small PRs, tackling one problem at a time. However, since `generate` is the main gate to LLMs in `transformers`, and caching is only needed for auto-regressive generation, day 0 support is important -- at least `gready_search` and `sample` should be operational by the time this PR can be merged. Otherwise, we might get ourselves in a position where we realize we need to rewrite a significant part of `Cache`/`generate` to enable them together, which is very undesirable. However, I'm pro having this PR enabling the new cache on a single model (Llama) and taking care of the other models later as needed 🤗 I can also give a hand on the `generate` side, when we are happy with the state of the PR with the exception of `generate`.

Re model-specific cache changes: I do expect some models to require different code to convert back and forth from the legacy cache format (not all models have the same cache format, there are ~5% of them have slight differences e.g. [see this in BLOOM](https://github.com/huggingface/transformers/blob/288bf5c1d2844e89a2a87aeee90033532335e2e6/src/transformers/models/bloom/modeling_bloom.py#L504)). You are also writing that RoPE-based models may also need custom logic. To me, that suggests five things:
1) a model may need to subclass the base `Cache` to implement its own nuances regarding the cache format
2) despite the above, the cache operations and output format are standardized for each type of cache, so we would benefit from a strong base class (as we do in the config files or in the base pretrained class).
3) I suspect the model-specific changes for the cache would be small, so we could keep them in the modeling file. We can always move to a new file if needed :)
4) Because of 1), each model would have to implement a thin wrapper class to the instantiable cache classes, even if the base class is fully compatible. This also goes in line with our philosophy in `tramsformers` where each model implements its own model-specific operations.
5) Because of 4), in practice, each model defines its own cache. This means users are free to write their own custom caches -- power to the users 💪 

The model-specific cache modification part is an important design decision that will have an impact over many versions of `transformers` to come, so I'd also like to hear your thoughts @tomaarsen @patrickvonplaten

gante commented:
Upon some shower thoughts, I've come across an alternative plan for the model-specific cache modification problem -- easier to implement and that would result in more readable code. 

Instead of the base `Cache` holding the code to convert to and from the legacy format (which then requires subclassing at each model, if it has a different legacy cache format), the conversion logic could be held in model-specific functions to convert to and from the legacy format. In other words, each model would implement a `to_legacy_cache` and a `from_legacy_cache`, and the different `Cache` implementations would be immediately available to the model with no further modifications!

patrickvonplaten commented:
> lding the code to convert to and from the legacy format (which then requires subclassing at each model, if it has a different legacy cache format)

Agree here! Think it would be better if we only have a few selected cache classes that work for all models. The functions `from_legacy_cache` and `to_legacy_cache` indeed need to be model specific, so I think they can be just stand-alone functions in each `modeling_....py` file. 

=> Thus, I think we can:
- a) Have all `Cache` classes in a general `generation/cache.py` file. These general cache implementations should be identical for all models
- b) For backwards compatibility each model needs specific `from_legacy_cache` and `to_legacy_cache` functions, but they don't need to be part of the Cache class IMO, they can just be stand-alone functions in each model class so that they are easier to deprecate going forward.

tomaarsen commented:
I just noticed that Joao's message has been edited, and I missed Patrick's message, so my response doesn't make much sense anymore - I deleted it.

I also prefer the `from_legacy_cache` and `to_legacy_cache` implementations. I'll get on it.

tomaarsen committed:
Implement the SinkCache through backward+forward rotations

tomaarsen committed:
Integrate (Sink)Cache with Llama FA2

tomaarsen committed:
Set use_legacy_cache=True as default, allows for test passes

tomaarsen committed:
Move from/to_legacy_cache to ...Model class

tomaarsen committed:
Undo unnecessary newline change

tomaarsen committed:
Remove copy utility from deprecated OpenLlama

tomaarsen committed:
Match import style

liangan1 commented:
@gante  @tomaarsen  This is really a good abstraction of kv_cache to enable window context and be compatible to the legacy kv_cache. But for the memory footprint, the 'torch.cat' are still needed when update the cache and reorder_cache using 'index_select' is also there with beam search.  
Do you have a plan to use the pre-allocate buffer to avoid 'torch.cat'?  The pre-allocated buffer should be compatible with semantic of your Cache.  For the cache.update operation,  token slots from the pre-allocated buffer are needed to store key/value token states.

tomaarsen commented:
I removed the commits regarding changing the order of the caches based on @gante's recommendation. He is working on more changes for this PR here: https://github.com/tomaarsen/transformers/pull/1

gante commented:
@liangan1 both are issues in the near-future roadmap, with this cache abstraction being a requirement! 🙌 

First we will work on pre-allocated buffers, which result in significant speedups (especially with `torch.compile`). Then we will simplify beam search into an XLA-friendly method, just like in our TF and JAX implementations.

liangan1 commented:
> @liangan1 both are issues in the near-future roadmap, with this cache abstraction being a requirement! 🙌
> 
> First we will work on pre-allocated buffers (a subclass of `Cache`), which result in significant speedups (especially with `torch.compile`). Then we will simplify beam search into an XLA-friendly method, just like in our TF and JAX implementations.

Wow, cool. we are also pleasure to do contributions. can you share more details about your roadmap?

patrickvonplaten commented on src/transformers/cache_utils.py:
```diff
@@ -0,0 +1,157 @@
+T = TypeVar("T")
+
+
+class Cache(ABC):
```
```suggestion
class Cache(ABC):
```

(nit) Do we really need the ABC abstraction? I don't think this does anything and don't see the reason to add it.

tomaarsen replied:
We don't need it, no. The reason I've added it is to avoid people from ever trying to initialize `Cache` rather than `DynamicCache`, but this can just be avoided by not importing `Cache` into `src/transformers/__init__.py`, i.e. preventing `from tranformers import Cache` from working.

@gante do you want to implement this in your work in https://github.com/tomaarsen/transformers/pull/1?

patrickvonplaten commented on src/transformers/cache_utils.py:
```diff
@@ -0,0 +1,157 @@
+        cos: Optional[torch.Tensor] = None,
+        sin: Optional[torch.Tensor] = None,
+    ) -> Tuple[torch.Tensor, torch.Tensor]:
+        pass
```
```suggestion
        raise NotImplementedError("Make sure to implement `update` in a child class.")
```

tomaarsen replied:
👍 except I would use "subclass" rather than "child class".

patrickvonplaten commented on src/transformers/cache_utils.py:
```diff
@@ -0,0 +1,157 @@
+            return 0
+        return self.key_cache[layer_idx].shape[-2]
+
+    def to_legacy_cache(self) -> Tuple[Tuple[torch.Tensor], Tuple[torch.Tensor]]:
```
Nice!

patrickvonplaten commented on src/transformers/cache_utils.py:
```diff
@@ -0,0 +1,157 @@
+        )
+
+    @classmethod
+    def from_legacy_cache(cls, past_key_values: Optional[List[torch.FloatTensor]]) -> "DynamicCache":
```
Nice!

patrickvonplaten reviewed (commented):
Should we roll out the new cache feature for the most important models and aim to get this PR in soon ?

liangan1 commented on src/transformers/models/llama/modeling_llama.py:
```diff
@@ -464,16 +466,12 @@ def forward(
-            kv_seq_len += past_key_value[0].shape[-2]
+            kv_seq_len += past_key_value.get_seq_length(self.layer_idx)
         cos, sin = self.rotary_emb(value_states, seq_len=kv_seq_len)
         query_states, key_states = apply_rotary_pos_emb(query_states, key_states, cos, sin, position_ids)
```
In the original code, the ROPE applied to the transposed query/key states, but here query/key is not transposed, experiment result show that this will cause the shape mismatch issue.

liangan1 commented on src/transformers/models/llama/modeling_llama.py:
```diff
@@ -464,16 +466,12 @@ def forward(
-            value_states = torch.cat([past_key_value[1], value_states], dim=2)
-
-        past_key_value = (key_states, value_states) if use_cache else None
+            key_states, value_states = past_key_value.update(key_states, value_states, self.layer_idx, cos, sin)
```
According to the Streaming LLM paper, the cache should store the original key/values w/o applied ROPE, but "key_states" here has been applied to ROPE, does it make sense?

liangan1 replied:
Thanks. That's clear enough.

gante committed:
manual rebase with main

gante committed:
Cache class working with generate (#1)

* Draft version of new KV Caching

This should allow Attention Sinks (https://github.com/tomaarsen/attention_sinks)
/ StreamingLLM (https://arxiv.org/abs/2309.17453) to be easily implemented
in a third-party or in transformers directly

* Address numerous PR suggestions

1. Move layer_idx from cache to ...Attention. Removes confusing set_layer_idx magic.
2. Always convert past_key_values to Cache instance at the start of ...Attention, removes all other isinstance calls.
3. Remove __bool__ and __getitem__ magic as they're confusing.
4. past_key_values.update(key, value, idx) now returns key, value.
5. Add use_legacy_cache flag, defaults to None, i.e. Falsey. This breaks generate for now, until 1) the cache is used is generate() or 2) use_legacy_cache is defaulted to True in generate() until we change it in another PR.
6. Separate key_cache and value_cache.

Some work is still needed to see if the SinkCache can conveniently be implemented with just one update method.

* Integrate (Sink)Cache with Llama FA2

* Move from/to_legacy_cache to ...Model class

* Undo unnecessary newline change

* Match import style

* working generate

* Add tests; Simplify code; Apply changes to Mistral and Persimmon

* fix rebase mess

* a few more manual fixes

* last manual fix

* propagate changes to phi

* upgrade test

* add use_legacy_cache docstring; beef up tests

* reintroduce unwanted deletes

---------

Co-authored-by: Tom Aarsen <Cubiegamedev@gmail.com>

gante commented:
This PR is now generate-compatible, and has tests to ensure the two formats are interchangeable 💪 

The suggested review order is `cache_utils.py` -> `modeling_llama.py` -> `test_utils.py` -> others

@patrickvonplaten in the added test, we confirm that the two formats are interchangeable: they can be converted back and forth and be manipulated the same way. However, under the hood, the updated models use the new cache internally. It would be best to add a long slow test to key models (Llama and Mistral here) before merging this PR, correct?

@tomaarsen can you confirm that the attention sinks are working as expected?

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_26681). All of your documentation changes will be reflected on that endpoint.

gante committed:
move import

gante committed:
add default to model_kwargs.get('use_legacy_cache')

gante committed:
correct failing test

gante commented on tests/test_modeling_common.py:
```diff
@@ -551,10 +551,6 @@ def check_training_gradient_checkpointing(self, gradient_checkpointing_kwargs=No
             return
 
         for model_class in self.all_model_classes:
-            config, inputs_dict = self.model_tester.prepare_config_and_inputs_for_common()
```
This was redundant with L565 (from the unmodified file). The lines below, such as `config.use_cache = False`, had no effect since `config` was overwritten in L565

patrickvonplaten commented on src/transformers/cache_utils.py:
```diff
@@ -0,0 +1,179 @@
+
+
+class Cache:
+    def __init__(self) -> None:
```
I'd be more in favor of moving all of the following methods:
`__init__`
`__getitem__`
`__iter__`
`__len__`
`get_seq_length`
`to_legacy_cache`
`from_legacy_cache`

 to `DynamicCache` 

because:
- a) We only need backward comp for the dynamic cache and IMO should restrict future Cache implementations to inherit these methods - that would be confusing for new users
- b) IMO `Cache` should a purely abstract object - i.e. it should only have abstract methods that raise `NotImplementedError(...)`
- c) SinkCache should not inherit methods that it then needs to overwrite and call with `super()` such as `get_seq_length` => this is not very readable and makes life hard for people that want to implement their own cache.
- d) I don't think every cache will have the following a list of tuples as `self.key_cache` - when we add the compiled cache we want to have a purely tensor-based key and value cache

gante replied:
Makes sense 👍 

It does add a few extra steps in the transition. `generate` will need a few touches (e.g. cache manipulation in beam search), but we can add them along the way.

In the long run, it is indeed better if we go on this route!

patrickvonplaten commented on src/transformers/generation/utils.py:
```diff
@@ -3263,6 +3264,8 @@ def beam_search(
             )
             if model_kwargs["past_key_values"] is not None:
                 model_kwargs["past_key_values"] = self._reorder_cache(model_kwargs["past_key_values"], beam_idx)
+                if not model_kwargs.get("use_legacy_cache", True):
```
```suggestion
                if not isinstance(past_key_values, Cache):
```
that's better IMO as it means we don't rely on the `use_legacy_cache` model_kwarg in case people forget to add it. The logic of "always convert to a `Cache` object when we don't have one should hold".

patrickvonplaten commented on src/transformers/generation/utils.py:
```diff
@@ -3598,6 +3601,8 @@ def beam_sample(
             )
             if model_kwargs["past_key_values"] is not None:
                 model_kwargs["past_key_values"] = self._reorder_cache(model_kwargs["past_key_values"], beam_idx)
+                if not model_kwargs.get("use_legacy_cache", True):
```
```suggestion
                if not isinstance(past_key_values, Cache):
```

patrickvonplaten commented on src/transformers/generation/utils.py:
```diff
@@ -3985,6 +3990,8 @@ def group_beam_search(
                 model_kwargs["past_key_values"] = self._reorder_cache(
                     model_kwargs["past_key_values"], reordering_indices
                 )
+                if not model_kwargs.get("use_legacy_cache", True):
```
```suggestion
                if not isinstance(past_key_values, Cache):
```

patrickvonplaten commented on src/transformers/generation/utils.py:
```diff
@@ -4325,6 +4332,8 @@ def constrained_beam_search(
             )
             if model_kwargs["past_key_values"] is not None:
                 model_kwargs["past_key_values"] = self._reorder_cache(model_kwargs["past_key_values"], beam_idx)
+                if not model_kwargs.get("use_legacy_cache", True):
```
```suggestion
                if not isinstance(past_key_values, Cache):
```

patrickvonplaten commented on src/transformers/generation/utils.py:
```diff
@@ -24,6 +24,7 @@
 import torch.distributed as dist
 from torch import nn
 
+from ..cache_utils import DynamicCache
```
```suggestion
from ..cache_utils import DynamicCache, Cache
```

patrickvonplaten commented on src/transformers/models/llama/modeling_llama.py:
```diff
@@ -278,9 +279,10 @@ def repeat_kv(hidden_states: torch.Tensor, n_rep: int) -> torch.Tensor:
     """Multi-headed attention from 'Attention Is All You Need' paper"""
 
-    def __init__(self, config: LlamaConfig):
+    def __init__(self, config: LlamaConfig, layer_idx: int):
```
```suggestion
    def __init__(self, config: LlamaConfig, layer_idx: Optional[int] = None):
```
Let's make sure we don't break backwards comp

patrickvonplaten replied:
But on the other hand we really do need the layer index for the caching mechanism now
Just to be a bit safer and make it easier for users to transition, can we maybe do the following:
- https://github.com/huggingface/transformers/pull/26681/files#r1402008864
- https://github.com/huggingface/transformers/pull/26681/files#r1402007386

patrickvonplaten commented on src/transformers/models/llama/modeling_llama.py:
```diff
@@ -378,16 +380,12 @@ def forward(
-            value_states = torch.cat([past_key_value[1], value_states], dim=2)
-
-        past_key_value = (key_states, value_states) if use_cache else None
+            key_states, value_states = past_key_value.update(key_states, value_states, self.layer_idx, cos, sin)
```
Nice! I like the idea of passing sin and cos here -> Are we sure that this doesn't lead to slow downs though?

tomaarsen replied:
It should not, see: https://github.com/huggingface/transformers/pull/26681/files/e4c072d53f9b275581c9e20dd1ce474e74bbf391#r1392564863

But I will try to incorporate some naive latency tests in my performance experiments today/tomorrow.

patrickvonplaten commented on src/transformers/models/llama/modeling_llama.py:
```diff
@@ -378,16 +380,12 @@ def forward(
         kv_seq_len = key_states.shape[-2]
         if past_key_value is not None:
-            kv_seq_len += past_key_value[0].shape[-2]
+            kv_seq_len += past_key_value.get_seq_length(self.layer_idx)
```
```suggestion
             if self.layer_idx is None:
                 raise ValueError(f"The cache structure has changed since version XXX. If you are using {self.__class__.__name__} for auto-regressive decoding with k/v caching, please make sure to initialize the attention class with a layer index.")
            kv_seq_len += past_key_value.get_seq_length(self.layer_idx)
```

patrickvonplaten commented on src/transformers/models/llama/modeling_llama.py:
```diff
@@ -278,9 +279,10 @@ def repeat_kv(hidden_states: torch.Tensor, n_rep: int) -> torch.Tensor:
+    def __init__(self, config: LlamaConfig, layer_idx: int):
         super().__init__()
         self.config = config
+        self.layer_idx = layer_idx
```
```suggestion
        self.layer_idx = layer_idx
        
        if layer_idx is None:
            logger.warn("Instantiating {self.__class__.__name__} without passing `layer_idx` is not recommended and can lead to errors during the forward call. Please make sure to provide a `layer_idx` when creating this class.")
```

gante replied:
Added `warning_once` instead, to avoid a giant list of repeated warnings :)

patrickvonplaten reviewed (commented):
The changes inside the modeling files look good, I think we should try hard though to minimize even backwards breaking changes for the attention classes API (or at least give nice warnings / errors).

I feel quite strongly though that the `Cache` object needs to be fully abstract with 0 functionality.

gante committed:
Apply suggestions from code review

Co-authored-by: Patrick von Platen <patrick.v.platen@gmail.com>

gante committed:
apply PR suggestions

gante committed:
fix failing test

gante commented:
@patrickvonplaten applied the suggested changes 👍  I agree they the full abstraction will be better in the long run.

Two minor modifications over your suggestions:
1. Replaced the `__init__` `warn` by `warning_once`, otherwise the user would get `n_layers` repeated warnings
2. In the beam methods, to check against `Cache`, we needed to perform the check before calling `_reorder_cache`

tomaarsen reviewed (commented):
First of all, some small nitpicks with the warnings.

Then, Attention Sinks works for Llama and Mistral in this PR, but **not** for Persimmon and Phi. This is because they both use partial rotary embeddings, e.g. for [Persimmon](https://github.com/huggingface/transformers/blob/1ddc4fa60e8cbf91367e33053361201dbffc193e/src/transformers/models/persimmon/modeling_persimmon.py#L285-L299) and for [Phi](https://github.com/huggingface/transformers/blob/1ddc4fa60e8cbf91367e33053361201dbffc193e/src/transformers/models/phi/modeling_phi.py#L303-L317).
The SinkCache always uses ["regular" rotary embeddings](https://github.com/tomaarsen/transformers/blob/bd91e13d252491d12c08eb50c626d45912b93089/src/transformers/cache_utils.py#L167), which won't work for these newer architectures. Perhaps each model architecture should be given the opportunity to "override" the default rotary function with some function that takes `key_states`, `cos` and `sin`?

Lastly, the SinkCache implementation here seems as efficient as my https://github.com/tomaarsen/attention_sinks implementation, i.e. the same memory usage and latency.

ArthurZucker commented:
Will review tomorrow !

ArthurZucker commented on src/transformers/cache_utils.py:
```diff
@@ -0,0 +1,183 @@
+        return cache
+
+
+def rotate_half(x):
+    """Rotates half the hidden dims of the input."""
+    x1 = x[..., : x.shape[-1] // 2]
+    x2 = x[..., x.shape[-1] // 2 :]
+    return torch.cat((-x2, x1), dim=-1)
+
+
+def apply_rotary_pos_emb_single(
+    key_states: torch.Tensor, cos: torch.Tensor, sin: torch.Tensor, position_ids: Optional[torch.IntTensor] = None
+) -> torch.Tensor:
+    if position_ids:
+        cos = cos[position_ids].unsqueeze(1)  # [seq_len, dim] -> [batch_size, 1, seq_len, head_dim]
+        sin = sin[position_ids].unsqueeze(1)
+    rotated_key_states = (key_states * cos) + (rotate_half(key_states) * sin)
+    return rotated_key_states
```
Don't think these should be here. They should be in the `SinkCache` class as we only need it there.

ArthurZucker commented on src/transformers/models/llama/modeling_llama.py:
```diff
@@ -278,9 +279,17 @@ def repeat_kv(hidden_states: torch.Tensor, n_rep: int) -> torch.Tensor:
+        self.layer_idx = layer_idx
+        if layer_idx is None:
+            logger.warning_once(
+                f"Instantiating {self.__class__.__name__} without passing `layer_idx` is not recommended and will "
+                "to errors during the forward call, if caching is used. Please make sure to provide a `layer_idx` "
+                "when creating this class."
```
```suggestion
                f"Instantiating {self.__class__.__name__} without passing `layer_idx` is not recommended and will "
                " lead to errors during the forward call if caching is used. Please make sure to provide a `layer_idx` "
                "when creating this class."
```

ArthurZucker commented on src/transformers/generation/utils.py:
```diff
@@ -3262,7 +3263,10 @@ def beam_search(
                 outputs, model_kwargs, is_encoder_decoder=self.config.is_encoder_decoder
             )
             if model_kwargs["past_key_values"] is not None:
+                using_new_cache = isinstance(model_kwargs["past_key_values"], Cache)
                 model_kwargs["past_key_values"] = self._reorder_cache(model_kwargs["past_key_values"], beam_idx)
+                if using_new_cache:
+                    model_kwargs["past_key_values"] = DynamicCache.from_legacy_cache(model_kwargs["past_key_values"])
```
`DynamicCache` is an instance of `Cache`, self._reorder_cache` returns 
```suggestion
                using_new_cache = isinstance(model_kwargs["past_key_values"], DynamicCache)
                model_kwargs["past_key_values"] = self._reorder_cache(model_kwargs["past_key_values"], beam_idx)
                if using_new_cache:
                    model_kwargs["past_key_values"] = DynamicCache.from_legacy_cache(model_kwargs["past_key_values"])
```
If we were using a `SinkCache` then the new cache's type will be lost + only the `DynamicCache` will have the `from_legacy_cache` no?

patrickvonplaten replied:
Same comment as: https://github.com/huggingface/transformers/pull/26681/files#r1412782842

ArthurZucker commented on src/transformers/models/llama/modeling_llama.py:
```diff
@@ -802,6 +808,9 @@ def _init_weights(self, module):
             more detail.
         return_dict (`bool`, *optional*):
             Whether or not to return a [`~utils.ModelOutput`] instead of a plain tuple.
+        use_legacy_cache (`bool`, *optional*):
+            If set to `True` (default), will return `past_key_values` as described input above. Otherwise, will return
+            a subclass of `Cache`
```
```suggestion
        use_legacy_cache (`bool`, *optional*):
            If set to `True` (default), will return `past_key_values` as a tuple, which is what we call the `legacy` cache format. Otherwise, will return a subclass of [`~cache_utils.Cache`]
```

ArthurZucker reviewed (commented):
Let's add a `cache.md` specific page to document the design choice and philosophy behind adding new caches! Otherwise it's in a good form, have a few comments about always having to pass cos and sin, and the `use_legacy_cache`.
Really sorry if I missed things explained offline 😓

patrickvonplaten commented on src/transformers/cache_utils.py:
```diff
@@ -0,0 +1,183 @@
+
+import torch
+
+
+class Cache:
```
(nit) Why use `ABC` here? :sweat_smile: I'm really not a fan of the ABC package as it adds zero functionality and for people that don't know the package it adds unnecessary logic that needs to be looked up.

When I see a class that has only functions that return `NotImplementedError` and a docstring that says "This is an abstract class" that's all I need and better than having a `ABC` abstraction

ArthurZucker replied:
I have zero strong opinion on ABC, let's use `NotImplementedError` and a good docstring then!

patrickvonplaten commented on src/transformers/cache_utils.py:
```diff
@@ -0,0 +1,183 @@
+        key_states: torch.Tensor,
+        value_states: torch.Tensor,
+        layer_idx: int,
+        cos: Optional[torch.Tensor] = None,
+        sin: Optional[torch.Tensor] = None,
```
```suggestion
        cos: Optional[torch.Tensor] = None,    # needed for models having RoPE position embeddings
        sin: Optional[torch.Tensor] = None,    # needed for models having RoPE position embeddings
```

patrickvonplaten commented on src/transformers/cache_utils.py:
```diff
@@ -0,0 +1,183 @@
+    return rotated_key_states
+
+
+class SinkCache(Cache):
```
I'd leave as is, I'd maybe just add some comments that `cos` / `sin` are only needed because of RoPE

patrickvonplaten commented on src/transformers/cache_utils.py:
```diff
@@ -0,0 +1,183 @@
+        key_states: torch.Tensor,
+        value_states: torch.Tensor,
+        layer_idx: int,
+        cos: Optional[torch.Tensor] = None,
+        sin: Optional[torch.Tensor] = None,
```
```suggestion
        pos_embeddings_dict: Optional[str, torch.Tensor]
```

Generally, I do agree that `cos`, `sin` is quite unique to RoPE embeddings. Another option would be to just add a general `pos_embeddings_dict: Optional[str, torch.Tensor]` to prevent the signature of `self.update` to explode.

patrickvonplaten replied:
But passing dicts is also not ideal for `torch.compile` or `torch.trace.jit`

gante committed:
Apply suggestions from code review

Co-authored-by: Patrick von Platen <patrick.v.platen@gmail.com>
Co-authored-by: Tom Aarsen <37621491+tomaarsen@users.noreply.github.com>

gante commented on src/transformers/models/llama/modeling_llama.py:
```diff
@@ -945,7 +957,9 @@ def forward(
             all_hidden_states += (hidden_states,)
 
-        next_cache = next_decoder_cache if use_cache else None
+        next_cache = None
+        if use_cache:
+            next_cache = next_decoder_cache.to_legacy_cache() if use_legacy_cache else next_decoder_cache
```
@ArthurZucker It behaves exactly like the old cache in the situations stipulated within `transformers`, but it is a new object type -- therefore technically a breaking change 😢 

I'm more than happy to proceed without this flag and jump straight into the new format. We can release it in v4.36 and see if anyone complains 👀 

LMK if you agree, in which case I will remove this flag.

tomaarsen replied:
cc: I'd like to hear @patrickvonplaten's opinion as well, as he proposed the flag.

Also, thanks a bunch @gante for picking up this PR from me.

patrickvonplaten replied:
I'm in favor of keeping the flag and slowly deprecating the tuple cache - don't see a big problem with having a flag

gante commented:
@ArthurZucker depending on your answer to [this comment](https://github.com/huggingface/transformers/pull/26681#discussion_r1411079517), the PR will move very differently -- I'll wait for your answer before continuing :)

patrickvonplaten commented on src/transformers/generation/utils.py:
```diff
@@ -3225,7 +3226,10 @@ def beam_search(
             if model_kwargs["past_key_values"] is not None:
+                using_new_cache = isinstance(model_kwargs["past_key_values"], Cache)
                 model_kwargs["past_key_values"] = self._reorder_cache(model_kwargs["past_key_values"], beam_idx)
+                if using_new_cache:
```
```suggestion
                if not isinstance(model_kwargs["past_key_values"], Cache):
```
Shouldn't it be like this?

patrickvonplaten replied:
Ah it's because `self._reorder_cache(...)` return a tuple? I'd be quite heavily in favor of adding a `_reorder_cache` to the `Cache` object. Then because in `generate` we only pass around objects of type `Cache` we can delete `_reorder_cache` completely no?

patrickvonplaten commented on src/transformers/models/llama/modeling_llama.py:
```diff
@@ -378,16 +387,18 @@ def forward(
-            value_states = torch.cat([past_key_value[1], value_states], dim=2)
-
-        past_key_value = (key_states, value_states) if use_cache else None
+            key_states, value_states = past_key_value.update(key_states, value_states, self.layer_idx, cos, sin)
```
We will very soon implement a static k/v cache here (as shown in https://github.com/pytorch-labs/gpt-fast). To implement a k/v cache, we need to mask future (not-yet-attended) attention mask tokens so that they don't influence the result. 

This means we also need to pass the `attention_mask` to the `update(...)` function. We could solve this by making `update(...)` accept a dict instead of just tensors: https://github.com/huggingface/transformers/pull/26681/files#r1405895192

gante committed:
PR comments

gante committed:
tmp commit

gante committed:
add docstrings

gante committed:
more tests, more docstrings, add to docs

gante commented:
Ready for a new round of reviews. IMO, it's now much better than before: no new flags, `generate` compatibility for any `Cache` subclass, among others.

@patrickvonplaten @ArthurZucker:
- There is no more `use_legacy_cache` flag. If `past_key_values` is a `Cache` instance (which can be an empty cache), the model sticks to that format. No flags to deprecate in the future :)
- Because of the point above, if we pass  `past_key_values=some_cache_instance` to `generate`, it will use that cache format. `generate` is ready to receive static caches.
- The caches now accept a `cache_kwargs` dictionary in `update` (its is not a `**dict`). It is the responsibility of each subclass to process this dictionary. It should gives us plenty of room (and flexibility) to grow without exploding the signature.
- `Cache` has a `reorder_cache` method, as suggested. This also means we can start thinking about standardizing cache formats and removing per-model functions;
- Added tons of docstrings

@tomaarsen:
- Added support for partial rotation in `SinkCache`. An extra set of eyes there would be appreciated! :)

⚠️ There is a bug with beam search + legacy cache format. I suspect it is a subtle dimension-related bug, but the PR can be reviewed while I chase it.

gante committed:
derp

patrickvonplaten reviewed (approved):
Very nice! Looks good to me

gante committed:
tmp commit

gante committed:
tmp dbg

gante committed:
more dbg

gante committed:
fix beam search bug

gante committed:
cache can be a list of tuples in some models

gante committed:
fix group beam search

tomaarsen reviewed (commented):
Just ran some tests on Persimmon with this, ran into a handful of issues, see the comments. It seems to work again after applying these changes.

cc: @gante

tomaarsen commented:
I've ran perplexity tests for Attention Sinks with Persimmon, Llama and Mistral, and they all look good in terms of performance, VRAM and latency. I plan to do some more tests with Phi tomorrow.
I'll also try to use SinkCache with `generate`.

gante committed:
all but sinkcache integration tests

gante committed:
fix sink cache and add hard integration test

gante committed:
now also compatible with input_embeds input

patrickvonplaten commented:
Let's get this one in for the release no? cc @ArthurZucker @amyeroberts @LysandreJik

ArthurZucker commented:
Yep

ArthurZucker commented on src/transformers/cache_utils.py:
```diff
@@ -0,0 +1,313 @@
+        else:
+            # Shifting cache
+
+            # # Incompatibility checks
+            # if key_states.shape[-2] > 1:
+            #     raise ValueError(
+            #         "Got past key values for more than one token at the shifting cache stage. `SinkCache` is not yet "
+            #         "compatible with techniques that generate more than one token at a time, like assisted "
+            #         "generation. Please open a GH issue if you see this issue with generation techniques that only "
+            #         "generate one token at a time!"
+            #     )
```
should probably be removed?

ArthurZucker commented on src/transformers/cache_utils.py:
```diff
@@ -0,0 +1,313 @@
+                [
+                    self.value_cache[layer_idx][:, :, : self.num_sink_tokens],
+                    self.value_cache[layer_idx][
+                        :, :, -self.window_length + self.num_sink_tokens + value_states.shape[-2] :
```
let's define the shape outside to make it more readble

ArthurZucker commented on src/transformers/models/llama/modeling_llama.py:
```diff
@@ -487,18 +499,13 @@ def forward(
-            value_states = torch.cat([past_key_value[1], value_states], dim=2)
-
-        past_key_value = (key_states, value_states) if use_cache else None
+            key_states, value_states = past_key_value.update(key_states, value_states, self.layer_idx, cos, sin)
```
```suggestion
            key_states, value_states = past_key_value.update(key_states, value_states, self.layer_idx, cache_kwargs)
```

ArthurZucker commented on src/transformers/models/llama/modeling_llama.py:
```diff
@@ -885,8 +901,11 @@ def forward(
-        if past_key_values is not None:
-            past_key_values_length = past_key_values[0][0].shape[2]
+        if use_cache:
+            use_legacy_cache = not isinstance(past_key_values, Cache)
```
thanks!

ArthurZucker commented on src/transformers/models/llama/modeling_llama.py:
```diff
@@ -1101,16 +1119,28 @@ def prepare_inputs_for_generation(
+
+            # If the cache has seen more tokens than it can hold, then the cache has a size limit. Let's discard the
+            # older attention values, as their corresponding values are not part of the input.
+            if cache_length < past_length and attention_mask is not None:
+                attention_mask = attention_mask[:, -(cache_length + input_ids.shape[1]) :]
```
that's specific to the sink cache no?

gante replied:
It will be for the fixed-length cache as well :D And windowed attention can probably make use of it as well

ArthurZucker replied:
alright :)

ArthurZucker commented on tests/generation/test_utils.py:
```diff
@@ -1904,6 +1906,66 @@ def test_generate_continue_from_past_key_values(self):
                     )
 
+    @parameterized.expand([(1, False), (1, True), (4, False)])
+    def test_new_cache_format(self, num_beams, do_sample):
```
nice test!

ArthurZucker commented on tests/test_cache_utils.py:
```diff
@@ -0,0 +1,189 @@
+        inputs = tokenizer(["Vaswani et al. (2017) introduced the Transformers"], return_tensors="pt").to(model.device)
+
+        # Set up the SinkCache. Using a small window length to contain computational complexity. If this example is run
+        # without a DynamicCache, the last few tokens are gibberish (ends in "of the of the of a of a of")
```
```suggestion
        # without a `SinkCache`, the last few tokens are gibberish (ends in "of the of the of a of a of")
```

ArthurZucker reviewed (approved):
Looks good to me, no additional kwargs, no generate back and forth conversion thanks for the great effort!

gante committed:
PR comments

gante committed:
add Cache support to Phi+FA2

gante committed:
make fixup

ydshieh commented:
I am going to trigger the CI now (with a reabse on main in order to include the fix #27887

ydshieh commented:
https://github.com/huggingface/transformers/actions/runs/7129508785

fxmarty commented on src/transformers/models/llama/modeling_llama.py:
```diff
@@ -889,8 +906,11 @@ def forward(
         past_key_values_length = 0
-        if past_key_values is not None:
-            past_key_values_length = past_key_values[0][0].shape[2]
+        if use_cache:
+            use_legacy_cache = not isinstance(past_key_values, Cache)
+            if use_legacy_cache:
+                past_key_values = DynamicCache.from_legacy_cache(past_key_values)
+            past_key_values_length = past_key_values.get_seq_length()
```
```suggestion
        if use_cache:
            use_legacy_cache = not isinstance(past_key_values, Cache) and past_key_values is not None
            if use_legacy_cache:
                past_key_values = DynamicCache.from_legacy_cache(past_key_values)
            elif past_key_values is None:
                past_key_values = DynamicCache()
            past_key_values_length = past_key_values.get_seq_length()
```

This has probably been discussed already but here I am: here in the first forward pass when using `generate`, we have `past_key_values` being None, and `use_legacy_cache` evaluates to `True`. This results in `from_legacy_cache` and `to_legacy_cache` being called at each decoding iteration. Is this intended for now?

Alternatively `DynamicCache()` may be initialized in `generation/utils.py` in case we care only about `generate` using the new cache implementation.

it's not clear to me why, but the proposed suggestion I give actually makes generation slower. Any idea what's happening?

fxmarty commented on src/transformers/models/llama/modeling_llama.py:
```diff
@@ -889,8 +906,11 @@ def forward(
         past_key_values_length = 0
-        if past_key_values is not None:
-            past_key_values_length = past_key_values[0][0].shape[2]
+        if use_cache:
+            use_legacy_cache = not isinstance(past_key_values, Cache)
+            if use_legacy_cache:
+                past_key_values = DynamicCache.from_legacy_cache(past_key_values)
+            past_key_values_length = past_key_values.get_seq_length()
```
```suggestion
        if use_cache:
            use_legacy_cache = not isinstance(past_key_values, Cache) and past_key_values is not None
            if use_legacy_cache:
                past_key_values = DynamicCache.from_legacy_cache(past_key_values)
            elif past_key_values is None:
                past_key_values = DynamicCache()
            past_key_values_length = past_key_values.get_seq_length()
```

This has probably been discussed already but here I am: here in the first forward pass when using `generate`, we have `past_key_values` being None, and `use_legacy_cache` evaluates to `True`. This results in `from_legacy_cache` and `to_legacy_cache` being called at each decoding iteration. Is this intended for now?

Alternatively `DynamicCache()` may be initialized in `generation/utils.py` in case we care only about `generate` using the new cache implementation.

it's not clear to me why, but the proposed suggestion I give actually makes generation slower. Any idea what's happening?

gante replied:
Hey @fxmarty! Glad to hear that it is compatible with the optimum ONNX export 💛 

Defaulting to the old cache type is intentional: for all intents and purposes, if we were to use (and return) a `DynamicCache` by default it would technically be an API change, so it has to go through a deprecation cycle. However, since caching touches many types of models and is used in lots of places, I want users to experiment with the feature and report bugs before placing the warning (like you just did with optimum ONNX export!). For instance, I'm fairly confident that we need code changes for the cache to support encoder-decoder models -- it has no room for the cross-attention atm. 

As for your suggestion being slower: no clue 👀 The conversion is pretty fast though, no tensors are created, so it should be cheap to leave it as is for a couple of minor `transformers` versions

fxmarty reviewed (commented):
Really awesome work. I'm hopeful we'll be able to get static cache thanks to this! Testing a bit this looks compatible with Optimum ONNX export which is cool!

fxmarty commented on src/transformers/cache_utils.py:
```diff
@@ -0,0 +1,298 @@
+    """
+
+    def __init__(self) -> None:
+        self.key_cache: List[torch.Tensor] = []
+        self.value_cache: List[torch.Tensor] = []
```
Why not initialize them as a list of empty tensors, or better, just a tensor (`(n_layers, batch_size, num_heads, 0, head_dim)`)?

They may allow to avoid
```python
        if len(self.key_cache) <= layer_idx:
            self.key_cache.append(key_states)
            self.value_cache.append(value_states)
```

fxmarty commented:
Actually it did break the ONNX export :D

stas00 commented on src/transformers/models/llama/modeling_llama.py:
```diff
@@ -283,9 +284,17 @@ def repeat_kv(hidden_states: torch.Tensor, n_rep: int) -> torch.Tensor:
+        self.layer_idx = layer_idx
+        if layer_idx is None:
+            logger.warning_once(
+                f"Instantiating {self.__class__.__name__} without passing `layer_idx` is not recommended and will "
+                "to errors during the forward call, if caching is used. Please make sure to provide a `layer_idx` "
+                "when creating this class."
```
this warning needs some TLC - the current "will to error" doesn't parse. Thank you!

tomaarsen replied:
You're right! I'll take care of it. Thanks @stas00

gante replied:
@tomaarsen FYI adding this fix to the next cache PR (adding support to encoder-decoder models), no need to open a PR :)

fxmarty commented on src/transformers/models/persimmon/modeling_persimmon.py:
```diff
@@ -802,16 +828,28 @@ def prepare_inputs_for_generation(
+            # 1 - If the length of the attention_mask exceeds the length of input_ids, then we are in a setting where
+            # some of the inputs are exclusivelly passed as part of the cache (e.g. when passing input_embeds as
+            # input)
+            if attention_mask is not None and attention_mask.shape[1] > input_ids.shape[1]:
+                input_ids = input_ids[:, -(attention_mask.shape[1] - past_length) :]
```
@tomaarsen When does this case happen?

fayejf commented:
@tomaarsen Great work! Looks like the current implementation cannot handle case **input_length > window_length** at least for some models (mistral) 
Do I miss anything? Thanks! 

```
cache = SinkCache(window_length=1024, num_sink_tokens=4)
gen_out = model.generate(**inputs, do_sample=False, max_new_tokens=10, 
                         use_cache=True,
                         past_key_values=cache, 
                         pad_token_id=tokenizer.eos_token_id)
decoded = tokenizer.batch_decode(gen_out, skip_special_tokens=True)
```
Supporse inputs is of torch.Size([1, 15759]). It will raise
`RuntimeError: shape '[-1, 15759]' is invalid for input of size 1024`
when it trys to reshape position_ids

```
File ~/anaconda3/envs/llm/lib/python3.8/site-packages/transformers/models/mistral/modeling_mistral.py:984, in MistralModel.forward(self, input_ids, attention_mask, position_ids, past_key_values, inputs_embeds, use_cache, output_attentions, output_hidden_states, return_dict)
    982     position_ids = position_ids.unsqueeze(0).view(-1, seq_length)
    983 else:
--> 984     position_ids = position_ids.view(-1, seq_length).long()
```

ArthurZucker commented:
Not sure it was tested fro mistral! Can you open a separate issue with a full reproducer? 🤗

MoonRide303 commented:
@fayejf You might want to look at this paper, too - handling extremely long inputs in a streaming fashion: [Leave No Context Behind: Efficient Infinite Context Transformers with Infini-attention](https://arxiv.org/abs/2404.07143).

> 1B model that was fine-tuned on up to 5K sequence length passkey instances solved the 1M length problem.

explanare commented:
@ArthurZucker It seems that even with Llama2, passing in SinkCache to `generate` causes errors.

I'm using transformers 4.39.3 and the Llama2 model was loaded using the following code:

```
from transformers import AutoConfig, LlamaForCausalLM, AutoTokenizer

model_id = "meta-llama/Llama-2-7b-hf"
model = LlamaForCausalLM.from_pretrained(
      model_id, low_cpu_mem_usage=True, device_map='auto',
      torch_dtype=torch.bfloat16)
model = model.eval()
tokenizer = AutoTokenizer.from_pretrained(model_id)
tokenizer.pad_token = tokenizer.eos_token
tokenizer.padding_side = 'left'
```

The SinkCache was passed to generate as in @fayejf's script. I am not sure if this is the correct way to use SinkCache:

```
from transformers import SinkCache

prefix = 'Hello world!'
inputs = tokenizer(prefix, return_tensors='pt').to(device)

cache = SinkCache(window_length=1024, num_sink_tokens=4)
gen_out = model.generate(**inputs, do_sample=False, max_new_tokens=256,
                         use_cache=True,
                         past_key_values=cache,
                         pad_token_id=tokenizer.pad_token_id)
decoded = tokenizer.batch_decode(gen_out, skip_special_tokens=True)
```

The code caused `TypeError: object of type 'SinkCache' has no len()` as a result of this `DynamicCache.from_legacy_cache` call (see the stack trace below). Looks like you are familiar with the `StaticCache` stuff, any suggestions on how to get around this? Thanks in advance!

```
File /miniconda3/envs/pytorch2/lib/python3.10/site-packages/transformers/models/llama/modeling_llama.py:977, in LlamaModel.forward(self, input_ids, attention_mask, position_ids, past_key_values, inputs_embeds, use_cache, output_attentions, output_hidden_states, return_dict, cache_position)
    975 if use_cache:  # kept for BC (cache positions)
    976     if not isinstance(past_key_values, StaticCache):
--> 977         past_key_values = DynamicCache.from_legacy_cache(past_key_values)
    978         past_seen_tokens = past_key_values.get_seq_length()
    980 if cache_position is None:

File /miniconda3/envs/pytorch2/lib/python3.10/site-packages/transformers/cache_utils.py:181, in DynamicCache.from_legacy_cache(cls, past_key_values)
    179 cache = cls()
    180 if past_key_values is not None:
--> 181     for layer_idx in range(len(past_key_values)):
    182         key_states, value_states = past_key_values[layer_idx]
    183         cache.update(key_states, value_states, layer_idx)
```

fayejf commented:
> @fayejf You might want to look at this paper, too - handling extremely long inputs in a streaming fashion: [Leave No Context Behind: Efficient Infinite Context Transformers with Infini-attention](https://arxiv.org/abs/2404.07143).
> 
> > 1B model that was fine-tuned on up to 5K sequence length passkey instances solved the 1M length problem.

@MoonRide303 Thanks for sharing! very interesting paper!

fayejf commented:
@ArthurZucker I was with 4.38.2 and haven't got a chance to test it with latest release. But looks like @explanare have seen similar issues. I also tried with llama model and I've seen legacy cache error as well. 

I'm not sure if I'm using SinkCache correctly. I didn't find any doc for it unfortuantely. :(
I follow the way in test file.

ys-zong commented:
> @ArthurZucker It seems that even with Llama2, passing in SinkCache to `generate` causes errors.
> 
> I'm using transformers 4.39.3 and the Llama2 model was loaded using the following code:
> 
> ```
> from transformers import AutoConfig, LlamaForCausalLM, AutoTokenizer
> 
> model_id = "meta-llama/Llama-2-7b-hf"
> model = LlamaForCausalLM.from_pretrained(
>       model_id, low_cpu_mem_usage=True, device_map='auto',
>       torch_dtype=torch.bfloat16)
> model = model.eval()
> tokenizer = AutoTokenizer.from_pretrained(model_id)
> tokenizer.pad_token = tokenizer.eos_token
> tokenizer.padding_side = 'left'
> ```
> 
> The SinkCache was passed to generate as in @fayejf's script. I am not sure if this is the correct way to use SinkCache:
> 
> ```
> from transformers import SinkCache
> 
> prefix = 'Hello world!'
> inputs = tokenizer(prefix, return_tensors='pt').to(device)
> 
> cache = SinkCache(window_length=1024, num_sink_tokens=4)
> gen_out = model.generate(**inputs, do_sample=False, max_new_tokens=256,
>                          use_cache=True,
>                          past_key_values=cache,
>                          pad_token_id=tokenizer.pad_token_id)
> decoded = tokenizer.batch_decode(gen_out, skip_special_tokens=True)
> ```
> 
> The code caused `TypeError: object of type 'SinkCache' has no len()` as a result of this `DynamicCache.from_legacy_cache` call (see the stack trace below). Looks like you are familiar with the `StaticCache` stuff, any suggestions on how to get around this? Thanks in advance!
> 
> ```
> File /miniconda3/envs/pytorch2/lib/python3.10/site-packages/transformers/models/llama/modeling_llama.py:977, in LlamaModel.forward(self, input_ids, attention_mask, position_ids, past_key_values, inputs_embeds, use_cache, output_attentions, output_hidden_states, return_dict, cache_position)
>     975 if use_cache:  # kept for BC (cache positions)
>     976     if not isinstance(past_key_values, StaticCache):
> --> 977         past_key_values = DynamicCache.from_legacy_cache(past_key_values)
>     978         past_seen_tokens = past_key_values.get_seq_length()
>     980 if cache_position is None:
> 
> File /miniconda3/envs/pytorch2/lib/python3.10/site-packages/transformers/cache_utils.py:181, in DynamicCache.from_legacy_cache(cls, past_key_values)
>     179 cache = cls()
>     180 if past_key_values is not None:
> --> 181     for layer_idx in range(len(past_key_values)):
>     182         key_states, value_states = past_key_values[layer_idx]
>     183         cache.update(key_states, value_states, layer_idx)
> ```

I got the same error with Llama2 using 4.38.2 following code [here](https://colab.research.google.com/drive/1S0oIPaqxAVp0oWEwTadhZXDjhWiTyF12?usp=sharing#scrollTo=fwxG9OFAV4Qv).
```
File "/opt/conda/envs/dn/lib/python3.10/site-packages/transformers/models/llama/modeling_llama.py", line 982, in forward
    past_key_values = DynamicCache.from_legacy_cache(past_key_values)
  File "/opt/conda/envs/dn/lib/python3.10/site-packages/transformers/cache_utils.py", line 166, in from_legacy_cache
    for layer_idx in range(len(past_key_values)):
TypeError: object of type 'SinkCache' has no len()
```

ArthurZucker commented:
cc @gante as well as we broke it

ArthurZucker commented:
Will might be fixed by #30476. Only tuples should be going that path

gante commented:
@fayejf @ys-zong @ArthurZucker 

`SinkCache` is operational on `main` with Mistral and a few other models (especially non-RoPE models). Llama and Llama-dependent models + `SinkCache` is fixed in [this PR](https://github.com/huggingface/transformers/pull/30581), which hopefully will get merged soon 🤗 

Apologies for this temporary issue, we had to break a few eggs (sink cache) to make an omelet (torch.compile support) :D

delock commented:
Hi @tomaarsen I just came across this line https://github.com/huggingface/transformers/blob/main/src/transformers/cache_utils.py#L979 in your PR and wonder why the condition for cache growing is  `key_states.shape[-2] + self.get_seq_length(layer_idx) < self.window_length`?  To my understanding even if the conditon is `<=` the new key states can still safely grow into the window size without need to evict.  Is there any logic that the codition must be `<`?  Thanks!
